runtime/dart_isolate.cc

```diff
@@ -0,0 +1,685 @@
+             package_config, flags, parent_embedder_isolate, error)
+      .first;
+}
+
+std::pair<Dart_Isolate, fml::WeakPtr<DartIsolate>>
+DartIsolate::CreateDartVMAndEmbedderObjectPair(
+    const char* advisory_script_uri,
+    const char* advisory_script_entrypoint,
+    const char* package_root,
+    const char* package_config,
+    Dart_IsolateFlags* flags,
+    DartIsolate* parent_embedder_isolate,
+    char** error) {
+  TRACE_EVENT0("flutter", "DartIsolate::CreateDartVMAndEmbedderObjectPair");
+  if (parent_embedder_isolate == nullptr ||
+      parent_embedder_isolate->GetDartVM() == nullptr) {
+    *error =
+        strdup("Parent isolate did not have embedder specific callback data.");
+    FXL_DLOG(ERROR) << *error;
+    return {nullptr, {}};
+  }
+
+  const DartVM* vm = parent_embedder_isolate->GetDartVM();
+
+  // Create the native object on the embedder side. This object is deleted in
+  // the cleanup callback.
+  auto embedder_isolate =
+      new DartIsolate(vm,                                             //
+                      parent_embedder_isolate->GetTaskRunners(),      //
+                      parent_embedder_isolate->GetResourceContext(),  //
+                      parent_embedder_isolate->GetSkiaUnrefQueue(),   //
+                      advisory_script_uri,                            //
+                      advisory_script_entrypoint                      //
+      );
+
+  // Create the Dart VM isolate and give it the embedder object as the baton.
+  Dart_Isolate isolate =
+      vm->GetPlatformKernel() != nullptr
+          ? Dart_CreateIsolateFromKernel(advisory_script_uri,         //
+                                         advisory_script_entrypoint,  //
+                                         vm->GetPlatformKernel(),     //
+                                         flags,                       //
+                                         embedder_isolate,            //
+                                         error                        //
+                                         )
+          : Dart_CreateIsolate(
+                advisory_script_uri,                                       //
+                advisory_script_entrypoint,                                //
+                vm->GetIsolateSnapshot().GetData()->GetSnapshotPointer(),  //
+                vm->GetIsolateSnapshot().GetInstructionsIfPresent(),       //
+                flags,                                                     //
+                embedder_isolate,                                          //
+                error                                                      //
+            );
+
+  if (isolate == nullptr) {
+    FXL_DLOG(ERROR) << *error;
+    return {nullptr, {}};
+  }
+
+  if (!embedder_isolate->Initialize(isolate)) {
+    *error = strdup("Embedder could not initialize the Dart isolate.");
+    FXL_DLOG(ERROR) << *error;
+    return {nullptr, {}};
+  }
+
```

Comment:
Shouldn't one of the PrepareForRunningFromX happen here?

Reply:
That requires a snapshot which can be acquired later. We launch the isolate and prepare it in two separate phases.